### DISCUSSION TITLE
Exiting project maintenance mode

### DISCUSSION THREAD

**Discussion by tjex** [zk-org/zk#405] [General]
Hey all @zk-org/maintainers 

As per the [maintenance mode](https://github.com/zk-org/zk/discussions/371#discussioncomment-8007312) plan, it would seem like we're due to exit back into the world of releases 🎉

There is one bug report left, which is my own. But it's to do with `zk graph`, which is a bit of an 'edge' feature. I think it's acceptable to consider this as a non-critical bug, with regards to cutting a new release or not. 

All bug reports over at zk-nvim have also been closed.

We also just got [one new bug report](https://github.com/zk-org/zk/discussions/371#discussioncomment-8007312) about the date parsing issue that we have already fixed.

Are there any concerns or reasons for not looking at cutting `zk 0.14.1` as a next step?

It also times nicely to cut the release before sprucing up the docs, [as discussed](https://github.com/zk-org/zk/discussions/388).

**Comment by mickael-menu** [comment]
That's fine by me. I'll let someone else do the release process to make sure I wrote everything needed in the MAINTAINING.md. 

Thanks again @tjex for all your maintenance work!

**Comment by tjex** [comment]
Yup, I agree. It's good if we work through it ourselves. It'll be a good stress 
test of the process and documentation. Thanks for writing that up and regarding 
your thanks, my pleasure! Really happy to be involved.

**Comment by tjex** [comment]
@mickael-menu ... Am I completely blind / can't use a file search?.. I can't find the MAINTAINING.md doc anymore.

**Reply by mickael-menu** [reply]
My bad, it was in CONTRIBUTING 😅
https://github.com/zk-org/zk/blob/main/CONTRIBUTING.md#releasing-a-new-version

**Comment by mcDevnagh** [comment]
Thank you tjex for taking the initiative here! I also propose we elevate you to _owner_, as you have been spearheading much of the progress in zk

**Reply by mickael-menu** [reply]
I agree, I changed your role in the `zk-org` to owner @tjex. Thanks for all the work done so far!

**Reply by tjex** [reply]
a pleasure 🥳

**Comment by tjex** [comment]
Draft release is under way. Changelog is done. 
Re. builds. I think I can get access to an intel mac to build the binary for it. 

If not, do either of you have access to one @mcDevnagh, @jurica or @julio-lopez ?

Also note https://github.com/zk-org/zk-xcompile/issues/3, which we should address for this and subsequent releases.

**Reply by mcDevnagh** [reply]
I could get access to a Mac if needed. I've compiled Go code for [Darwin](https://stackoverflow.com/questions/65881808/how-can-i-cross-compile-to-darwin-386-from-linux) without a Mac before though

**Reply by tjex** [reply]
Hmm interesting. I was just going off what Mickael wrote in the contributing.md, that the mac intel builds need to be ran from intel. But I'll give it a crack, thanks :) Will ping you back if needed for the intel mac access.

I'm on a mac m1 myself.

**Reply by mickael-menu** [reply]
If I recall correctly, the issue is that using CGO prevents using the default cross-compilation provided in Go. See my comment on the contributing doc:

> This process is convoluted because zk requires CGO with mattn/go-sqlite3, which prevents using Go's native cross-compilation. Transitioning to a CGO-free SQLite driver such as [cznic/sqlite](https://gitlab.com/cznic/sqlite) could simplify the distribution process significantly.

**Reply by tjex** [reply]
So a little update:

- I've made my own images from zk-xcompile, by updgrading the go version to 1.21 from which they're based. 
- seemingly from go 1.19+ there is an error regarding building within docker images and [vsc stamping](https://github.com/golang/go/issues/51748#issuecomment-1072458099). So am currently investigating that. The easy fix is to just switch off VCS stamping, but I'm not sure of the implications of this.

If I switch VCS off: 

- linux arm is building fine
- i386 returns the error: 

```text
cgo: exec gcc: fork/exec /usr/bin/gcc: exec format error
```

- amd64 returns the error:

```text
# runtime/cgo
gcc: error: unrecognized command-line option '-m64'
```

Also getting these errors with the standard golang docker image on docker hub (i.e. `docker pull golang:1.21`).

And, still need to setup an environment on the intel mac I have around.

But it's in progress.

**Reply by tjex** [reply]
Builds are working now :) But haven't looked at mac intel build yet. Not expecting strife from that though. 

It came down to the docker file needing an explicit platform setting in FROM --platform=linux/amd64 golang:1.21 🤷‍♂️

And maybe also using docker buildx instead of just plain docker build to make the new images... Didn't test builds in between those two steps.

So just need to get the mac intel build and we're good.

Will update the contributing documentation and make a PR for zk-xcompile with the new images too. Hopefully over the weekend.

**Reply by mickael-menu** [reply]
Nice!

> But haven't looked at mac intel build yet. Not expecting strife from that though.

Yes, it's just a matter of running `make dist-macos` from an Intel Mac. It will produce a zip file like `zk-v0.14.0-macos-arm64.zip` if run from a tagged commit.

The Docker build (and needing an Intel Mac) for the release is really cumbersome. If you're serious in maintaining `zk` in the long term and adding more features, I would really recommend taking a look into getting rid of the CGO dependencies (with [cznic/sqlite](https://gitlab.com/cznic/sqlite) for example).

**Reply by tjex** [reply]
@zk-org/maintainers 

intel build is done. So just need to sort out what we're doing about #412 , which is connected to https://github.com/zk-org/zk-xcompile/pull/4 and linked issue https://github.com/zk-org/zk-xcompile/issues/3

Concluding those will define the proper and final state for the source code (i.e. the Makefile) to bundle with the release. 

Also waiting for feedback from community over on #411 about an Alpine build. It was fairly straight forward to spin up the [docker container to build for Alpine](https://github.com/tjex/zk-xcompile/tree/alpine), and figured it's in our best interests to release for Alpine as well. Are there any objections to that?

For that, https://github.com/zk-org/zk/pull/414 should also be approved and merged.

**Reply by jurica** [reply]
First off all I'm sorry, I was really busy in the past few weeks and haven't had any time to check what's going on here.
I'm afraid I don't own a Mac anymore, but I remember that go cross compilation with CGO on macOS wasn't really a smooth experience, but IIRC it is possible.
@tjex how did you build the release binaries and source tarballs. I looked into the Makefile and GitHub Actions but can't find anything there. Could you please point me in the right direction or aren't these build on GH Actions?

**Reply by tjex** [reply]
Not a prob 👍

The builds aren't happening via actions. We need to cross compile with docker (instead of using go's native cross compilation) because we're building with `CGO` (due to using mattn/go-sqlite3).

The build + tarball commands are [here in the makefile](https://github.com/zk-org/zk/blob/578894f45b0e8d117f29077fe5d376d6ad838eca/Makefile#L29)

Detail on that and more is in the [contributing doc](https://github.com/zk-org/zk/blob/main/CONTRIBUTING.md#releasing-a-new-version)

**Reply by jurica** [reply]
Ah I see, but we could still run this on actions as it's possible to run docker containers or am I missing something here?
I could take a look into this, IMHO having a working automated build is worth a lot.

**Reply by tjex** [reply]
I'm not experienced with running github builds, so if you think it's possible then that would be great :)
I think the issue then making a docker image for mac? I did find this project though:

https://hub.docker.com/r/sickcodes/docker-osx

**Reply by jurica** [reply]
I filed [an issue](https://github.com/zk-org/zk/issues/421#issue-2334150606) and outline what I have in mind therefor, then while I'm experimenting we can refine and discuss the goals.
In the past I have setup and maintained quite a lot of CI/CD pipelines so I hope I can also get this done ;).
As for the macOS part, I would just build on plain macOS machines, I don't think the hassle with docker for/on macOS is worth the effort.
